File: leetcode/binarySearch/binarySearch.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func main() {
	list := make([]int, 1000000)
	for i := 0; i < 1000000; i++ {
		list = append(list, i+1)
	}

	rand.Seed(time.Now().UnixNano()) //设置随机种子
	for i := 0; i < 20; i++ {
		v := rand.Intn(1000000 - 1) + 1 //生成1-1000000的随机数
		fmt.Printf("针对 %d 进行二分查找：\n", v)
		idex := binarySearch(list, v)
		fmt.Printf("%d 的索引位置是:[%d]\n",v, idex)
		fmt.Println("-------------------------------------")
	}
}

func binarySearch(list []int, target int) int {
	low := 0
	high := len(list) - 1

	step := 0 //记录查找次数

	for {
		step++
		if low <= high {
			mid := (low + high) / 2
			guess := list[mid]
			if guess == target {
				fmt.Printf("共查找了 %d 次\n", step)
				return mid
			}
			if guess < target {
				low = mid + 1
			} else {
				high = mid - 1
			}
		}
	}
}

File: leetcode/binarySearch/binarySearch.cpp
#include <iostream>
#include <cstdlib>
#include <time.h>
using namespace std;

int binarySearch(int list[], int target, int l, int h) {
    int low = l;
    int high = h;
    int step = 0; // 查找次数
    while (low <= high) {
        step ++;
        int mid = (low + high) / 2;
        if (list[mid] == target) {
            printf("查找次数为 %d\n", step);
            return mid;
        }
        if (list[mid] < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
}
// 数组作为函数参数传递的时候，传递的是指针，所以不能正确的求出子函数中传递进去的数组的长度，切记
int main() {
    int *list = new int[1000000];
    for (int i = 0; i < 1000000; i++) {
        list[i] = i;
    }
    int l = 0;
    int h = 100000; //数组元素个数
    srand(time(NULL)); // 随机数种子
    for (int i = 0; i < 20; i++) {
        int v = rand() % 1000000 + 1; //
        printf("查找的数字是%d", v);
        int index = binarySearch(list, v, l, h);
        printf("索引是%d\n", index);
    }
}
